interfaces/builtin/shared_memory_test.go

```diff
@@ -0,0 +1,292 @@
+	var sharedMemoryYaml = `name: consumer
+version: 0
+plugs:
+ shmem:
+  interface: shared-memory
+  %s
+apps:
+ app:
+  plugs: [shmem]
+`
+	data := []struct {
+		plugYaml      string
+		expectedError string
+	}{
+		{
+			"shared-memory: [one two]",
+			`shared-memory "shared-memory" attribute must be a string, not \[one two\]`,
+		},
+	}
+
+	for _, testData := range data {
+		snapYaml := fmt.Sprintf(sharedMemoryYaml, testData.plugYaml)
+		_, plug := MockConnectedPlug(c, snapYaml, nil, "shmem")
+		err := interfaces.BeforePreparePlug(s.iface, plug)
+		c.Check(err, ErrorMatches, testData.expectedError, Commentf("yaml: %s", testData.plugYaml))
+	}
+}
+
+func (s *SharedMemoryInterfaceSuite) TestPlugShmAttribute(c *C) {
+	var plugYamlTemplate = `name: consumer
+version: 0
+plugs:
+ shmem:
+  interface: shared-memory
+  %s
+apps:
+ app:
+  plugs: [shmem]
+`
+
+	data := []struct {
+		plugYaml     string
+		expectedName string
+	}{
+		{
+			"",      // missing "shared-memory" attribute
+			"shmem", // use the name of the plug
+		},
+		{
+			"shared-memory: shmemFoo",
+			"shmemFoo",
+		},
+	}
+
+	for _, testData := range data {
+		snapYaml := fmt.Sprintf(plugYamlTemplate, testData.plugYaml)
+		_, plug := MockConnectedPlug(c, snapYaml, nil, "shmem")
+		err := interfaces.BeforePreparePlug(s.iface, plug)
+		c.Assert(err, IsNil)
+		c.Check(plug.Attrs["shared-memory"], Equals, testData.expectedName,
+			Commentf(`yaml: %q`, testData.plugYaml))
+	}
+}
+
+func (s *SharedMemoryInterfaceSuite) TestSanitizeSlot(c *C) {
+	c.Assert(interfaces.BeforePrepareSlot(s.iface, s.slotInfo), IsNil)
+}
+
+func (s *SharedMemoryInterfaceSuite) TestSanitizeSlotUnhappy(c *C) {
+	var sharedMemoryYaml = `name: provider
+version: 0
+slots:
+ shmem:
+  interface: shared-memory
+  %s
+apps:
+ app:
+  slots: [shmem]
+`
+	data := []struct {
+		slotYaml      string
+		expectedError string
+	}{
+		{
+			"shared-memory: 12",
+			`shared-memory "shared-memory" attribute must be a string, not 12`,
+		},
+		{
+			"", // missing "write" attribute
+			`shared memory interface requires at least a valid "read" or "write" attribute`,
+		},
+		{
+			"write: a string",
+			`shared-memory "write" attribute must be a list of strings, not "a string"`,
+		},
+		{
+			"read: [Mixed, 12, False, list]",
+			`shared-memory "read" attribute must be a list of strings, not "\[Mixed 12 false list\]"`,
+		},
+		{
+			`read: ["ok", "trailing-space "]`,
+			`shared-memory interface path has leading or trailing spaces: "trailing-space "`,
+		},
+		{
+			`write: [" leading-space"]`,
+			`shared-memory interface path has leading or trailing spaces: " leading-space"`,
+		},
+		{
+			`write: [""]`,
+			`shared-memory interface path is empty`,
+		},
+		{
+			`write: [mem/**]`,
+			`shared-memory interface path is invalid: "mem/\*\*" contains a reserved apparmor char.*`,
+		},
+		{
+			`write: [/dev/shm/bar]`,
+			`shared-memory interface path should not contain '/': "/dev/shm/bar"`,
+		},
+		{
+			`write: [mem/../etc]`,
```

Comment:
could you add two more test cases, with both read and write where one is valid and the other is invalid?

Reply:
Added.